src/unk_02025534.c

```diff
@@ -1,33 +1,405 @@
+    GF_ASSERT(obj != NULL);
+    return obj->plttKey;
+}
+
+u32 G3dResFileHeader_GetSizeWithoutTex(const NNSG3dResFileHeader *header) {
+    return ResFileHeaderGetSizeWithoutTex(header);
+}
+
+static GF_3DGfxRawResObj *_3DGfxResMan_FindNextFreeObjSlot(const GF_3DGfxRawResMan *obj) {
+    for (int i = 0; i < obj->man->max; ++i) {
+        if (obj->objs[i].obj == NULL) {
+            return &obj->objs[i];
+        }
+    }
+
+    return NULL;
+}
+
+static void _3DGfxResObj_Init(GF_3DGfxRawResObj *obj) {
+    obj->obj = NULL;
+    obj->texKey = 0;
+    obj->tex4x4Key = 0;
+    obj->plttKey = 0;
+    obj->hasLoadedFromHeader = FALSE;
+    obj->resFileHeader = NULL;
+    obj->hasLoadedFromHeader = FALSE;  // BUG: should be headerNeedsLoad?
+}
+
+static NNSG3dResTex *_3DGfxResObj_GetTex_Internal(GF_3DGfxRawResObj *obj) {
+    return NNS_G3dGetTex(GF2dGfxRawResObj_GetData(obj->obj));
+}
+
+static NNSG3dResTex *_3DGfxResObj_GetTex_MaybeFromSecondaryHeader(GF_3DGfxRawResObj *obj) {
+    return NNS_G3dGetTex(!obj->headerNeedsLoad ? GF2dGfxRawResObj_GetData(obj->obj) : obj->resFileHeader);
+}
+
+static void ResTexAllocVramAndGetKeys(NNSG3dResTex *tex, NNSG3dTexKey *texKey, NNSG3dTexKey *tex4x4Key, NNSG3dPlttKey *plttKey) {
+    u32 texSize;
+    u32 tex4x4Size;
+    u32 plttSize;
+
+    texSize = NNS_G3dTexGetRequiredSize(tex);
+    tex4x4Size = NNS_G3dTex4x4GetRequiredSize(tex);
+    plttSize = NNS_G3dPlttGetRequiredSize(tex);
+
+    if (texSize != 0) {
+        *texKey = NNS_GfdAllocTexVram(texSize, FALSE, 0);
+    }
+    if (tex4x4Size != 0) {
+        *tex4x4Key = NNS_GfdAllocTexVram(tex4x4Size, TRUE, 0);
+    }
+    if (plttSize != 0) {
+        *plttKey = NNS_GfdAllocPlttVram(plttSize, tex->tex4x4Info.flag & NNS_G3D_RESPLTT_USEPLTT4, 0);
+    }
+}
+
+static void ResTexLoad(NNSG3dResTex *tex, GF_3DGfxRawResObj *obj) {
+    ResTexSetKeys(tex, obj->texKey, obj->tex4x4Key, obj->plttKey);
+    DC_FlushRange(tex, tex->header.size);
+    GF3dRender_LoadTexture(tex, TRUE);
+    GF3dRender_LoadPalette(tex, TRUE);
+}
+
+static void ResTexSetKeys(NNSG3dResTex *tex, NNSG3dTexKey texKey, NNSG3dTexKey tex4x4Key, NNSG3dPlttKey plttKey) {
+    NNS_G3dTexSetTexKey(tex, texKey, tex4x4Key);
+    NNS_G3dPlttSetPlttKey(tex, plttKey);
+}
+
+static void ResTexReleaseKeys(NNSG3dResTex *tex) {
+    NNSG3dTexKey texKey;
+    NNSG3dTexKey tex4x4Key;
+    NNS_G3dTexReleaseTexKey(tex, &texKey, &tex4x4Key);
+    NNS_G3dPlttReleasePlttKey(tex);
+}
+
+static void *ResFileHeaderCopyWithoutTex(const NNSG3dResFileHeader *header, HeapID heapId) {
+    u32 size = ResFileHeaderGetSizeWithoutTex(header);
+    void *ret = AllocFromHeap(heapId, size);
+    memcpy(ret, header, size);
+    return ret;
+}
+
+static u32 ResFileHeaderGetSizeWithoutTex(const NNSG3dResFileHeader *header) {
+    NNSG3dResTex *tex = NNS_G3dGetTex(header);
+    GF_ASSERT(tex != NULL);
+    return (const u8 *)NNS_G3dGetTexData(tex) - (const u8 *)header;
```

Comment:
why the casting? can't you just define a const u8* ret variable a line above to make it nicer or will it not match

Reply:
defining additional variables will not make it nicer so not worth pursuing